### ISSUE TITLE
Strategy for 0.6.0

### ISSUE THREAD

**Issue by rosskevin** [mdbootstrap/mdb-ui-kit#747]
> As discussed in https://github.com/FezVrasta/bootstrap-material-design/issues/733#issuecomment-159388374
> 
> Lazy loading variables in less makes it fundamentally different to the order of inclusion of variables for them to override.  Less requires your overrides last, Sass requires them prior to inclusion of what you would like to override.
> 
> BS4 is moving to sass, and BS3 has stable bower package `bootstrap-sass` maintained officially as part of BS3.  

@FezVrasta - I'm moving 0.6.0 to sass.

I recommend it becoming the master, with 0.5.0 living on with the dual less/sass version on a branch.  If someone wants to step up and maintain 0.5.0, we can accept PRs.

As of getting a stable 0.6.0 sass version, I have no plans to look back.

With a 0.6.0 that is stable, do you agree that should be merged to master?

**Comment by rosskevin** [comment]
@FezVrasta - we had an internal discussion and I've decided to get started on BS V4 alpha immediately.  So 0.6.0 as it is specified probably should turn into something like 4.0.0.  So I'll be getting a head start on the next major version, starting with our current sass code.

**Comment by Zhigalin** [comment]
@rosskevin 

> As of getting a stable 0.6.0 sass version, I have no plans to look back.

Your conception of "stability" is pretty strange, you know?
The master branch is not a place for raw alphas.
I mean that master should always be production-ready, so that someone can just run 

> git clone; cp dist/\* project/assets/ 

and go ahead, and update with

> git pull;  cp dist/\* project/assets/ 

without worries like

> Oh shi@!!!11 Everything is broken here!!!!!!111

Progress is a good thing, but there souldn't be an "always alpha" state, because only a framework that no one uses can have a master kinda like a nightly Ubuntu build. Look at Bootstrap guys, they said they'll continue to support v3 because dropping support of production-ready v2 in favour of v3 was really a bad idea.

**Comment by FezVrasta** [comment]
Our development branch is `master`, use the git tags to get the latest stable releases.

The same approach is used by Adobe with its Brackets editor project.

**Comment by rosskevin** [comment]
@Zhigalin - I'm not sure you read my last comment, but I'll address your comment.   

I think you have an entirely different conception of how to use git.

**First**: `master` is expected to be always in flux, a place for integration where development features are merged.  It is the tip of the repository and in _no_ way should be considered production ready.

Tags for releases should be considered stable, as should old version maintenance branches.

**Second**: This project to-date has always been an _alpha_, that's what I'm trying to improve
As stated on the README:

> NOTE: This theme is still in development, it could be used on production websites but I can't guarantee compatibility with previous versions.

This project has never been completed for BS V3 or the material spec, and I have been moving it forward rapidly, as well as responding rapidly to the point release bugs.  I have put time into making sure that each release has notes about compatibilities, even though this warning exists.  I'm trying to make it not just an alpha, but that means a lot of work and commits.

**Third**:  This is open source and each contributor needs to solve their own problems.  While a member can certainly make requests, each member cannot forget that the committers have their own needs and products to complete.  If a member needs something, they should expect to have to build it and hopefully submit a PR for the betterment of the community.  While submitting issues with test cases is very helpful, demanding their completion is not.  I have seen a bit of that in reviewing all of the issues here.  I am the only one committing right now, and I don't see many PR's.   

What I'm saying is - if you feel strongly about something - _great_ - welcome to open source, you are well suited to be a contributor.  Get started on your first PR!

**In conclusion**
We all benefit from progress and community.   Part of my effort here is to get this project to a state where:
1) it satisfies my needs of using material design with bootstrap
2) is perceived by the community to be worthwhile, adopted, and more maintainers step-up to reduce my maintenance load

@FezVrasta has clearly put in a lot of time, and I'm putting in my time to get it to the next milestone.

**Comment by Zhigalin** [comment]
@rosskevin 
First:
Master is a bleeding edge branch? Really?
Look at the Bootstrap itself.
The master is on v3.3.6 _not_ v4
It look like those twitter guys have

> an entirely different conception of how to use git.

Huh?

Second:

> This project to-date has always been an alpha

v0.4.4 has been pretty close to a release candidate, anyway no one was able to call it an alpha, but...

> that's what I'm trying to improve

v0.4.4 was almost production-ready, v0.5 is an alph... no, a _pre_alpha so this statement is just bogus.
I understand that you are trying to get the things done, but you can't just ruin the whole thing people rely on.

> As stated on the README

I have read it but I'd newer thought you guys are _that_ crazy and going to commit so many changes of _that_ dimension.

Third:

> This is open source

Oh yeah, I know.
This is the reason I hate opensouce for.

> Maybe we are going to fix the duchain? That bug will soon have his 4th birthday...
> Who cares? Lets rewrite the project in version 5!
> But we are going to lose all features from the current version!
> Who cares?

And then you can do nothing, and you are just looking as our common work is being send to /dev/null 
And then a mature product is an alpha again, once more time.
Others are excited because of new hipster design, new framework and modern technologies under the hood, and only few people looks sad like me. That happened again... Next time they will decide to rewrite this in CoffeeScript or maybe in Brainfuck... 

> This is open source

This is SPARTAAAA!!!!11

>  I am the only one committing right now, and I don't see many PR's. What I'm saying is - if you feel strongly about something - great - welcome to open source, you are well suited to be a contributor. Get started on your first PR!

I were just setting my dev environment and I really was ready to work hard but then I saw
https://github.com/FezVrasta/bootstrap-material-design/commit/828fd32669f8c8c478724fed6d86580faae3bcab
And then I said WTF.

> @FezVrasta has clearly put in a lot of time.

And a big part of his effort was deleted with a few commits.

To be clear: I'm not a troll, I'm just really sad because now I have to continue the v0.4.4 development(and consider that I am not a fronted man, I am a backed guy!) and I have just one pair of hands.
Seriously, guys, why you haven't just made a sandbox branch v4-dev like Boothtrap have done?

**Comment by rosskevin** [comment]
@Zhigalin I think you have misunderstood several things that are occurring.
1. Git methodology
   
   Master is the edge for the current release version (compatible with BS3), that methodology is not going to change.  When we are ready to move v4 to master, we will have a main branch for v3 compatible code and maintenance.  All codelines will be maintainable, I'm not sure why you see this as a problem.   Perhaps you think that for some reason the changes on the master are for v4 when they are not?
2.  Regarding https://github.com/FezVrasta/bootstrap-material-design/commit/828fd32669f8c8c478724fed6d86580faae3bcab
   
   I'm unclear of the problem, do you think I deleted color variation capabilities?  Fullpalette variations were problematic, and perhaps you don't understand or didn't review the final 0.5.0 tag (which is master), but the entire full palette remains supported, it just isn't in a single compiled massive css.  You just simply set your variables and compile.  I was actually very careful and _nothing_ to my knowledge was wiped out.  I did delete the shell script because it was virally licensed, incompatible with MIT, and did nothing more than would be achieved by changing variables and running `grunt dist`.  As far as I know, 0.5.0 is on master and is as stable as 0.4.x.
   
   So your flaming `WTF`, do you think I still deserve that?
3. Branching for v4-dev
   
   [`v4-dev`](https://github.com/FezVrasta/bootstrap-material-design/tree/v4-dev) was branched last night and I've been working on it all day.  It is sass and has a new documentation framework. It has nothing to do with 0.4.x or 0.5.x.
4. Constructive discourse
   
   Let's stick to constructive criticism without the color commentary and make this project better together.

**Comment by FezVrasta** [comment]
Just my 2 cents...

If you want to use this project is better for you to use [Bower](http://bower.io) or [NPM](https://www.npmjs.com/) to manage dependencies so you can correctly manage the versioning of them.

No one in 2015 uses `git clone; cp dist/* project/assets/` anymore.

**Comment by Zhigalin** [comment]
@FezVrasta I have an allergy on server-side javascript...

**Comment by ivantcholakov** [comment]
I take the sources manually and compile them with PHP scripts. :-)

**Comment by Zhigalin** [comment]
@ivantcholakov Wow, is nice to see a backend guy like me here :)

**Comment by rosskevin** [comment]
I am a back end guy.  A couple of tech generations now, at least coding for 15 years.   I use bower with rails.  It is far easier than manual management. 

Package managers have been prolific with back end developers since maven 1 with Java. Avoiding a simple package manager like bower or npm with today's css/javascript is like holding out for manual configuration of transitive dependencies in ant to win the community over maven. It isn't easier and the open communities have already chosen for us.  Resistance is futile,  I assure you that the hour of time it takes to learn and integrate it is worthwhile.

**Comment by FezVrasta** [comment]
In the company I work for we use bower and npm to manage dependencies, and we work strictly with the kernel and all the backend stuff. I think we are the most "backend guys" of the bay right now. :)

**Comment by rosskevin** [comment]
Funny that two back end guys have put the most commits into a design framework :)

**Comment by FezVrasta** [comment]
@rosskevin ahah I'm mostly frontend, but I'm one of the few at the company. All the time I hear people talking about java, kernels, docker and stuff like that.

**Comment by rosskevin** [comment]
It might be unfair to call me "back end" anymore.  Im actually an entrepreneur that retired and came back to start a product company because I love building things.   I'm more or less a jack of all trades,  though we have a partner that is visual only,  and one that is back end developer only,  I fill in every other gap.

**Comment by Zhigalin** [comment]
@rosskevin Show me please where I said that I'm against package managers?
I'm against server side JavaScript.

**Comment by FezVrasta** [comment]
Then you may use bower with PHP if you prefer.

**Comment by Zhigalin** [comment]
@FezVrasta OMG, thanks.
I have missed this one.
